app/services/teachers/school_transfers/history.rb

```diff
@@ -0,0 +1,72 @@
+module Teachers::SchoolTransfers
+  class History
+    ALLOWED_TEACHER_TYPES = %i[ect mentor].freeze
+
+    Transfer = Data.define(:leaving_training_period, :joining_training_period) do
+      def type
+        return :unknown unless joining_training_period
+        return :new_provider if leaving_training_period.lead_provider !=
+          joining_training_period.lead_provider
+
+        :new_school
+      end
+
+      def status
+        return :complete unless leaving_training_period.finished_on.future? ||
+          joining_training_period&.started_on&.future?
+
+        :incomplete
+      end
+
+      def teacher_type = leaving_training_period.for_ect? ? :ect : :mentor
+    end
+
+    def initialize(teacher:, lead_provider:, teacher_type:)
+      @teacher = teacher
+      @lead_provider = lead_provider
+
+      unless teacher_type.in?(ALLOWED_TEACHER_TYPES)
+        raise ArgumentError, "Invalid teacher type"
+      end
+
+      @teacher_type = teacher_type
+    end
+
+    def transfers
+      school_periods.map.with_index { |school_period, index|
+        next unless school_period.complete?
+
+        next_school_period = school_periods[index + 1]
+        next if next_school_period &&
+          (school_period.school == next_school_period.school)
+
+        leaving_training_period = school_period.latest_training_period
+        next unless leaving_training_period.complete?
+
```

Comment:
I think we're missing a scenario here where - if we're on the _last_ `SchoolPeriod` and `TrainingPeriod` and the `Teacher` has completed their training - it would not be classed as a leaving transfer (as they have finished their training). Sorry I should have had an explicit scenario for this, but its sort of the inverse of scenario 2 in the ticket.

Maybe check with Claire or Andy, but I think a teacher has 'completed training' when the `teacher.finished_induction_period&.complete?`